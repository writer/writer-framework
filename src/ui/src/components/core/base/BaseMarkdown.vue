<template>
	<BaseMarkdownRaw
		v-if="isMarkedLoaded"
		:raw-markdown="String(marked.parse(props.rawText)).trim()"
		class="BaseMarkdown"
	></BaseMarkdownRaw>
</template>

<script setup lang="ts">
import { onBeforeMount, ref } from "vue";
import BaseMarkdownRaw from "./BaseMarkdownRaw.vue";

let marked: typeof import("marked");
const isMarkedLoaded = ref(false);

const props = defineProps<{
	rawText: string;
}>();

onBeforeMount(async () => {
	marked = await import("marked");
	isMarkedLoaded.value = true;
});
</script>
