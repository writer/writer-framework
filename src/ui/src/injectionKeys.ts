import { ComputedRef, InjectionKey, Ref, VNode } from "vue";
import {
	BuilderManager,
	CollaborationManager,
	Component,
	Core,
	InstancePath,
	InstancePathItem,
	NotesManager,
	SecretsManager,
} from "./writerTypes";

export default {
	core: Symbol() as InjectionKey<Core>,
	builderManager: Symbol() as InjectionKey<BuilderManager>,
	notesManager: Symbol() as InjectionKey<NotesManager>,
	collaborationManager: Symbol() as InjectionKey<CollaborationManager>,
	secretsManager: Symbol() as InjectionKey<SecretsManager>,
	evaluatedFields: Symbol() as InjectionKey<Record<string, ComputedRef<any>>>,
	componentId: Symbol() as InjectionKey<Component["id"]>,
	isBeingEdited: Symbol() as InjectionKey<Ref<boolean>>,
	isDisabled: Symbol() as InjectionKey<Ref<boolean>>,
	getChildrenVNodes: Symbol() as InjectionKey<
		(instanceNumber?: InstancePathItem["instanceNumber"]) => VNode[]
	>,
	renderProxiedComponent: Symbol() as InjectionKey<
		(
			componentId: Component["id"],
			instanceNumber: InstancePathItem["instanceNumber"],
			ext?: { class?: string[]; contextSlot?: string },
		) => VNode
	>,
	instancePath: Symbol() as InjectionKey<InstancePath>,
	flattenedInstancePath: Symbol() as InjectionKey<string>,
	instanceData: Symbol() as InjectionKey<Ref[]>,
};
