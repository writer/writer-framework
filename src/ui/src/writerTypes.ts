import type { Component as VueComponent } from "vue";
import { generateCore } from "./core";
import { generateBuilderManager } from "./builder/builderManager";
import type { SchemaObject } from "ajv";

export type Core = ReturnType<typeof generateCore>;

type ComponentId = string;

/**
 * Basic building block of applications.
 * Multiple instances of a single component can exists. For example, via Repeater.
 */

type VisibleField = {
	expression: boolean | string; // True | False | 'custom'
	binding: string; // variable binding when expression is custom
	reversed: boolean;
};

export type Component = {
	id: ComponentId;
	parentId: string;
	type: string;
	position: number;
	content: Record<string, string>;
	isCodeManaged?: boolean;
	handlers?: Record<string, string>;
	visible?: VisibleField;
	binding?: {
		eventType: string;
		stateRef: string;
	};
	x?: number;
	y?: number;
	outs?: {
		outId: string;
		toNodeId: string;
	}[];
};

/**
 * Identifies a unique instance of a Component.
 */

export type InstancePathItem = {
	componentId: Component["id"];
	instanceNumber: number;
};

/**
 * Details the full path, including all ancestors, of a unique instance of a Component.
 */
export type InstancePath = InstancePathItem[];

/**
 * Defines component structure and behaviour. Included in Component templates.
 */
export type WriterComponentDefinitionField = {
	/** Display name */
	name: string;
	/** Initial value */
	init?: string;
	/** Description */
	desc?: string;
	/** Value used if the field is empty, e.g. "(No text)" */
	default?: string;
	/** Which control (text, textarea, etc) to use if not the default for the type */
	control?: FieldControl;
	options?:
		| Record<string, string>
		| ((wf?: Core, componentId?: ComponentId) => Record<string, string>); // List of values to be provided as autocomplete options
	/** Data type for the field */
	type: FieldType;
	/** Category (Layout, Content, etc) */
	category?: FieldCategory;
	/** Use the value of this field as a CSS variable */
	applyStyleVariable?: boolean;
	validator?: SchemaObject;
};

export type WriterComponentDefinition = {
	name: string; // Display name for the component
	description: string; // Short description
	docs?: string; // Collapsible mini-docs
	toolkit?: "core" | "workflows";
	category?: string; // Category (Layout, Content, etc)
	allowedChildrenTypes?: (string | "*" | "inherit")[]; // Which component types are allowed inside (if any)
	allowedParentTypes?: string[]; // Which component types can contain this type of component
	slot?: string; // In which slot component should render whgen "*" is used it will render in all slots
	fields?: Record<
		string, // Id for the field
		WriterComponentDefinitionField
	>;
	events?: Record<
		string, // Event type
		{
			desc?: string; // Description
			stub?: string; // Stub method for the event
			bindable?: boolean; // Whether this event is used for value binding
		}
	>;
	previewField?: string; // Which field to use for previewing in the Component Tree
	positionless?: boolean; // Whether this type of component is positionless (like Sidebar)
	outs?: Record<
		string,
		{
			name: string;
			description: string;
			style: string;
			field?: keyof WriterComponentDefinition["fields"];
		}
	>;
};

export type BuilderManager = ReturnType<typeof generateBuilderManager>;

export const enum ClipboardOperation {
	Cut = "cut",
	Copy = "copy",
}

export const enum FieldType {
	Text = "Text",
	KeyValue = "Key-Value",
	Color = "Color",
	Shadow = "Shadow",
	Number = "Number",
	Object = "Object",
	IdKey = "Identifying Key",
	Width = "Width",
	HAlign = "Align (H)",
	VAlign = "Align (V)",
	Padding = "Padding",
	Tools = "Tools",
}

export const enum FieldCategory {
	General = "General",
	Style = "Style",
	Tools = "Tools",
}

/**
 * Used to specify the field control if it's different from the default for the FieldType.
 */
export const enum FieldControl {
	Text = "Text",
	Textarea = "Textarea",
}

export type ComponentMap = Record<Component["id"], Component>;

/**
 * Unit of data for non-state-mutation communications between frontend and backend.
 */
export type MailItem = { type: string; payload: Record<string, string> };

export type UserFunction = { name: string; args: string[] };

export type AbstractTemplate = {
	baseType: string;
	writer: WriterComponentDefinition;
};

export type TemplateMap = Record<string, VueComponent>;

export type SourceFilesFile = {
	type: "file";
	complete?: boolean;
	content: string;
};

export type SourceFilesDirectory = {
	type: "directory";
	children: Record<string, SourceFiles>;
};

/**
 * Represent a file tree as an object with:
 *
 * - the key representing the filename
 * - the content as a string for a readable text file, or as a recursive `SourceFiles` if it's a directory
 *
 * @example
 * ```json
 * { "type": "directory", "children": { "main.py": { "type": "file", "content": "print('hello')", "complete": true } } }
 * ```
 */
export type SourceFiles = SourceFilesDirectory | SourceFilesFile;
