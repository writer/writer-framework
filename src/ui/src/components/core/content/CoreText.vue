<template>
	<div
		v-if="shouldDisplay"
		ref="rootEl"
		class="CoreText"
		:style="rootStyle"
		@click="handleClick"
	>
		<BaseEmptiness v-if="isEmpty" :component-id="componentId" />
		<template v-else>
			<BaseMarkdown
				v-if="fields.useMarkdown.value == 'yes'"
				:raw-text="fields.text.value"
				:style="contentStyle"
			>
			</BaseMarkdown>
			<div v-else class="plainText" :style="contentStyle">
				{{ fields.text.value }}
			</div>
		</template>
	</div>
</template>

<script lang="ts">
import { cssClasses, primaryTextColor } from "@/renderer/sharedStyleFields";
import { getClick } from "@/renderer/syntheticEvents";
import { FieldCategory, FieldControl, FieldType } from "@/writerTypes";

const clickHandlerStub = `
def click_handler(state):

	# Increment counter when the text is clicked

	state["counter"] += 1`;

const description =
	"A component to display plain text or formatted text using Markdown syntax.";

export default {
	writer: {
		name: "Text",
		description,
		category: "Content",
		fields: {
			text: {
				name: "Text",
				init: "Text",
				desc: "Add text directly, or reference state elements with @{my_text}.",
				type: FieldType.Text,
				control: FieldControl.Textarea,
			},
			useMarkdown: {
				name: "Use Markdown",
				desc: "The Markdown output will be sanitised; unsafe elements will be removed.",
				default: "no",
				type: FieldType.Text,
				options: {
					yes: "Yes",
					no: "No",
				},
			},
			alignment: {
				name: "Alignment",
				default: "left",
				type: FieldType.Text,
				options: {
					left: "Left",
					center: "Center",
					right: "Right",
				},
				category: FieldCategory.Style,
			},
			primaryTextColor,
			cssClasses,
		},
		events: {
			"wf-click": {
				desc: "Capture single clicks.",
				stub: clickHandlerStub.trim(),
			},
		},
		previewField: "text",
	},
};
</script>
<script setup lang="ts">
import { computed, inject, useTemplateRef } from "vue";
import injectionKeys from "@/injectionKeys";
import BaseEmptiness from "../base/BaseEmptiness.vue";
import BaseMarkdown from "../base/BaseMarkdown.vue";

const rootEl = useTemplateRef("rootEl");
const fields = inject(injectionKeys.evaluatedFields);
const componentId = inject(injectionKeys.componentId);
const wf = inject(injectionKeys.core);

const isBeingEdited = inject(injectionKeys.isBeingEdited);
const isEmpty = computed(() => !fields.text.value);
const shouldDisplay = computed(() => !isEmpty.value || isBeingEdited.value);

const rootStyle = computed(() => {
	const component = wf.getComponentById(componentId);
	const isClickHandled =
		typeof component.handlers?.["wf-click"] !== "undefined";

	return {
		cursor: isClickHandled ? "pointer" : "unset",
	};
});

const contentStyle = computed(() => {
	return {
		"text-align": fields.alignment.value,
	};
});

function handleClick(ev: MouseEvent) {
	const ssEv = getClick(ev);
	rootEl.value.dispatchEvent(ssEv);
}
</script>

<style scoped>
@import "@/renderer/sharedStyles.css";

.CoreText {
	color: var(--primaryTextColor);
	line-height: 140%;
	white-space: pre-wrap;
	max-width: 100%;
	overflow: hidden;
	font-size: 0.875rem;
	font-weight: 400;
}

.CoreText ol,
.CoreText ul {
	white-space: normal;
}

.CoreText img {
	width: 100%;
}
</style>
