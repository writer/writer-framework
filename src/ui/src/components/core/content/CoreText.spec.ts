import { describe, expect, it, beforeEach, beforeAll } from "vitest";
import BaseMarkdown from "../base/BaseMarkdown.vue";
import CoreText from "./CoreText.vue";
import injectionKeys from "@/injectionKeys";
import { buildMockComponent, buildMockCore, mockProvides } from "@/tests/mocks";
import { shallowMount } from "@vue/test-utils";
import { ref } from "vue";
import { Core } from "@/writerTypes";

describe("CoreText", async () => {
	const text = `# Hello\n\nI'm the **content**.`;
	let core: Core;

	beforeAll(() => {
		core = buildMockCore().core;

		core.addComponent(
			buildMockComponent({
				handlers: { "wf-click": "python_handler" },
			}),
		);
	});

	it("should render in non-markdown mode", () => {
		const wrapper = shallowMount(CoreText, {
			global: {
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.isBeingEdited as symbol]: ref(false),
					[injectionKeys.evaluatedFields as symbol]: {
						text: ref(text),
						alignment: ref("center"),
						useMarkdown: ref(false),
					},
				},
			},
		});

		expect(wrapper.element).toMatchSnapshot();
	});

	it("should render in markdown mode", () => {
		const wrapper = shallowMount(CoreText, {
			global: {
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.isBeingEdited as symbol]: ref(false),
					[injectionKeys.evaluatedFields as symbol]: {
						text: ref(text),
						alignment: ref("center"),
						useMarkdown: ref(true),
					},
				},
			},
		});

		expect(wrapper.findComponent(BaseMarkdown).exists()).toBe(true);

		expect(wrapper.element).toMatchSnapshot();
	});
});
