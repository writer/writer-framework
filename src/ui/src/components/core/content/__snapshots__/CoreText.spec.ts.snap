// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoreText > should render in markdown mode 1`] = `
<div
  class="CoreText"
  data-v-810f69c0=""
  style="cursor: pointer;"
>
  
  <base-markdown-stub
    data-v-810f69c0=""
    rawtext="# Hello

I'm the **content**."
    style="text-align: center;"
  />
  
</div>
`;

exports[`CoreText > should render in non-markdown mode 1`] = `
<div
  class="CoreText"
  data-v-810f69c0=""
  style="cursor: pointer;"
>
  
  <p
    class="plainText"
    data-v-810f69c0=""
    style="text-align: center;"
  >
    # Hello

I'm the **content**.
  </p>
  
</div>
`;
